File: bak/some_cocos_proj_srcs/NetPacket.cpp
#include "Precompile.h"


NetPacket::NetPacket() : _wo( 0 )
, _ro( 0 )
, _bufLen( (int)NetConfigs::PacketBufferSize - sizeof( Header ) )
{
    _buf = new char[ (int)NetConfigs::PacketBufferSize ] + sizeof( Header );
}

NetPacket::NetPacket( int maxPkgLen ) : _wo( 0 )
, _ro( 0 )
{
    if( maxPkgLen == 0 )
    {
        maxPkgLen = (int)NetConfigs::PacketBufferSize;
        _bufLen = (int)NetConfigs::PacketBufferSize - sizeof( Header );
    }
    else _bufLen = maxPkgLen - sizeof( Header );
    assert( maxPkgLen >= sizeof( Header ) );
    _buf = new char[ maxPkgLen ] + sizeof( Header );
}

NetPacket::NetPacket( NetPacket const & other )
: _ro( other._ro )
, _wo( other._wo )
, _bufLen( other._bufLen )
{
    _buf = new char[ other.bufLen() ];
    memcpy( _buf, other.bufData(), other.bufLen() );
}

NetPacket::NetPacket( NetPacket && other )
: _ro( other._ro )
, _wo( other._wo )
, _bufLen( other._bufLen )
, _buf( other._buf )
{
    other._buf = nullptr;
}

NetPacket & NetPacket::operator=( NetPacket const & other )
{
    _ro = other._ro;
    _wo = other._wo;
    _bufLen = other._bufLen;
    if( _buf )
        delete[] bufData();
    _buf = new char[ other.bufLen() ];
    memcpy( _buf, other.bufData(), other.bufLen() );
    return *this;
}

NetPacket & NetPacket::operator=( NetPacket && other )
{
    _ro = other._ro;
    _wo = other._wo;
    _bufLen = other._bufLen;
    swap( _buf, other._buf );
    return *this;
}

NetPacket::~NetPacket()
{
    if( _buf )
        delete[] bufData();
}

string NetPacket::dump()
{
    prepare();
    char buf[ 128 ];
    sprintf( buf, "Packet Dump\nHeader: Packet Length = %d, OpCode = %d\nBuffer: Data Buffer Length = %d, Data Length = %d"
             , pkgLen(), opcode(), _bufLen, _wo );
    string s = buf;
    BufferHelper::dump( s, pkgData(), pkgLen() );
    printf( "%s\n", s.c_str() );  // 立即输出， lua 可能收接不下
    return "";
}

void NetPacket::dwu4( uint32 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dwu2( uint16 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dwu1( uint8 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dw4( int32 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dw2( int16 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dw1( int8 v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dwf( float v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dwd( double v )
{
    BufferHelper::write( _buf, _wo, v );
}

void NetPacket::dws( const char * v )
{
    int len = (int)strlen( v ) + 1; BufferHelper::write( _buf, _wo, v, len );
}

uint32 NetPacket::dru4()
{
    uint32 v; BufferHelper::read( v, _buf, _ro ); return v;
}

uint16 NetPacket::dru2()
{
    uint16 v; BufferHelper::read( v, _buf, _ro ); return v;
}

uint8 NetPacket::dru1()
{
    uint8 v; BufferHelper::read( v, _buf, _ro ); return v;
}

int32 NetPacket::dr4()
{
    int32 v; BufferHelper::read( v, _buf, _ro ); return v;
}

int16 NetPacket::dr2()
{
    int16 v; BufferHelper::read( v, _buf, _ro ); return v;
}

int8 NetPacket::dr1()
{
    int8 v; BufferHelper::read( v, _buf, _ro ); return v;
}

float NetPacket::drf()
{
    float v; BufferHelper::read( v, _buf, _ro ); return v;
}

double NetPacket::drd()
{
    double v; BufferHelper::read( v, _buf, _ro ); return v;
}

string NetPacket::drs()
{
    string v; BufferHelper::read( v, _buf, _ro ); return v;
}

void NetPacket::clearRo()
{
    _ro = 0;
}

void NetPacket::clearWo()
{
    _wo = 0;
}

void NetPacket::clear()
{
    _ro = _wo = 0;
}

int NetPacket::owfree( int offset ) const
{
    return _bufLen - offset;
}

int NetPacket::orfree( int offset ) const
{
    return _wo - offset;
}

int NetPacket::wfree() const
{
    return _bufLen - _wo;
}

int NetPacket::rfree() const
{
    return _wo - _ro;
}

int NetPacket::roffset() const
{
    return _ro;
}

void NetPacket::roffset( int v )
{
    _ro = v;
}

int NetPacket::woffset() const
{
    return _wo;
}

void NetPacket::woffset( int v )
{
    _wo = v;
}

uint16 NetPacket::opcode() const
{
    return header().opcode;
}

void NetPacket::opcode( uint16 v )
{
    header().opcode = v;
}

int NetPacket::bufLen() const
{
    return _bufLen + sizeof( Header );
}

char * NetPacket::bufData() const
{
    return _buf - sizeof( Header );
}

int NetPacket::pkgLen() const
{
    return _wo + sizeof( Header );
}

char * NetPacket::pkgData() const
{
    return _buf - sizeof( Header );
}

char * NetPacket::data() const
{
    return _buf;
}

int NetPacket::dataLen() const
{
    return _wo;
}

void NetPacket::prepare()
{
    header().pkgLen = pkgLen();
}

NetPacket* NetPacket::create( int maxPkgLen /*= 0 */ )
{
    auto rtv = new (nothrow) NetPacket( maxPkgLen );
    if( !rtv ) return nullptr;
    rtv->autorelease();
    return rtv;
}

File: bak/some_cocos_proj_srcs/api/NetPacket.lua

--------------------------------
-- @module NetPacket
-- @extend Ref
-- @parent_module ww

--------------------------------
-- @function [parent=#NetPacket] dump 
-- @param self
-- @return string#string ret (return value: string)
        
--------------------------------
-- @function [parent=#NetPacket] orfree 
-- @param self
-- @param #int int
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] clear 
-- @param self
        
--------------------------------
-- @function [parent=#NetPacket] wfree 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] dws 
-- @param self
-- @param #char char
        
--------------------------------
-- @function [parent=#NetPacket] prepare 
-- @param self
        
--------------------------------
-- @function [parent=#NetPacket] dr1 
-- @param self
-- @return char#char ret (return value: char)
        
--------------------------------
-- @function [parent=#NetPacket] bufLen 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] dr2 
-- @param self
-- @return short#short ret (return value: short)
        
--------------------------------
-- @function [parent=#NetPacket] dr4 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] owfree 
-- @param self
-- @param #int int
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] dwf 
-- @param self
-- @param #float float
        
--------------------------------
-- @function [parent=#NetPacket] clearRo 
-- @param self
        
--------------------------------
-- @function [parent=#NetPacket] dwd 
-- @param self
-- @param #double double
        
--------------------------------
-- @function [parent=#NetPacket] dru1 
-- @param self
-- @return unsigned char#unsigned char ret (return value: unsigned char)
        
--------------------------------
-- @overload self, int         
-- @overload self         
-- @function [parent=#NetPacket] woffset
-- @param self
-- @param #int int

--------------------------------
-- @function [parent=#NetPacket] dru2 
-- @param self
-- @return unsigned short#unsigned short ret (return value: unsigned short)
        
--------------------------------
-- @function [parent=#NetPacket] dru4 
-- @param self
-- @return unsigned int#unsigned int ret (return value: unsigned int)
        
--------------------------------
-- @function [parent=#NetPacket] pkgLen 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] dwu1 
-- @param self
-- @param #unsigned char char
        
--------------------------------
-- @function [parent=#NetPacket] dwu2 
-- @param self
-- @param #unsigned short short
        
--------------------------------
-- @function [parent=#NetPacket] dwu4 
-- @param self
-- @param #unsigned int int
        
--------------------------------
-- @function [parent=#NetPacket] dw2 
-- @param self
-- @param #short short
        
--------------------------------
-- @function [parent=#NetPacket] dw1 
-- @param self
-- @param #char char
        
--------------------------------
-- @function [parent=#NetPacket] dw4 
-- @param self
-- @param #int int
        
--------------------------------
-- @function [parent=#NetPacket] clearWo 
-- @param self
        
--------------------------------
-- @function [parent=#NetPacket] rfree 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- @function [parent=#NetPacket] drs 
-- @param self
-- @return string#string ret (return value: string)
        
--------------------------------
-- @overload self, int         
-- @overload self         
-- @function [parent=#NetPacket] roffset
-- @param self
-- @param #int int

--------------------------------
-- @overload self, unsigned short         
-- @overload self         
-- @function [parent=#NetPacket] opcode
-- @param self
-- @param #unsigned short short

--------------------------------
-- @function [parent=#NetPacket] drd 
-- @param self
-- @return double#double ret (return value: double)
        
--------------------------------
-- @function [parent=#NetPacket] drf 
-- @param self
-- @return float#float ret (return value: float)
        
--------------------------------
-- @function [parent=#NetPacket] create 
-- @param self
-- @return NetPacket#NetPacket ret (return value: NetPacket)
        
return nil
